Add support for Arc and Sector primitive

aurelj opened the pull request:
Hi! Thank you for helping out with Embedded Graphics development! Please:

- [x] Check that you've added passing tests and documentation
- [x] Add a simulator example(s) where applicable
- [x] Add a `CHANGELOG.md` entry in the **Unreleased** section under the appropriate heading (**Added**, **Fixed**, etc) and appropriate crate (`embedded-graphics`, `simulator`, `tinytga`, `tinybmp`) if your changes affect the **public API**
- [x] Run `rustfmt` on the project
- [x] Run `./build.sh` (Linux/macOS only) and make sure it passes. If you use Windows, check that CI passes once you've opened the PR.

## PR description

This is a first draft implementation of the Arc primitive.
For now it supports filled Arcs, but I plan to implement Sector and remove fill support from Arc, as described in #284. I think Arc and Sector will use a common `StyledIterator` but I'm not sure yet how to do this cleanly.
There is probably no need for a full detailed review right now, but I would like to know if this is overall going into the right direction.

Here are a few technical details of the current implementation:
- Angles are specified as integer degrees (this is plenty enough for me right now, and it shouldn't be very hard to add support for radiants)
- Center + Angles are converted to linear equation using `tan()`
- `tan()` is implemented with a LUT of `FixedPoint` values
- I introduced a basic implementation of `FixedPoint` numbers

jamwaffles commented:
I had a (very) quick scroll and this stood out:

1. I would prefer to support the `(center: Point, start: Point, end: Point)` definition first for my immediate use case (joiners and end caps for polylines) over the current implementation using angles. This seems at odds with your requirements, so it would be great if this PR supported both.
2. That's a lot of lookup table, and a lot of code for `FixedPoint`. If we do have to use that approach, is there an existing crate we can use? I'm moderately strongly against adding this functionality into e-g.
3. I haven't thought about this too deeply, but is it possible to use Bresenham-like integer maths to replace the `LinearEquation` structs? I haven't taken the time to understand the algo so this could be a misguided comment.
4. Are all the tiny `combine_*` functions required?

aurelj commented:
> I had a (very) quick scroll and this stood out:
> 
> 1. I would prefer to support the `(center: Point, start: Point, end: Point)` definition first for my immediate use case (joiners and end caps for polylines) over the current implementation using angles. This seems at odds with your requirements, so it would be great if this PR supported both.

Indeed, we don't have the same need at all. I understand your use case and I agree that it would be great to support both. I will see what I can do.
But I have an issue with your `(center: Point, start: Point, end: Point)` definition. It looks over-constrained to me. What if the distance between `center` and `start` is not the same as the distance between `center` and `stop` ? We couldn't draw a Circle or Arc that passes thru both `start` and `stop`.
Maybe the definition should be `(start: Point, end: Point, radius: u32)` with a way to specify on which side to position the center (it could be for example at the right of the `start` -> `end` vector).
I also guess that for this use case you will always want the minor Arc and never the major one.

> 2. That's a lot of lookup table, and a lot of code for `FixedPoint`. If we do have to use that approach, is there an existing crate we can use? I'm moderately strongly against adding this functionality into e-g.

Maybe we could use the `fixed` crate to replace my `FixedPoint` implementation. I will have a look at it.
But we will still need our `tan()` implementation (ie. LUT).

> 3. I haven't thought about this too deeply, but is it possible to use Bresenham-like integer maths to replace the `LinearEquation` structs? I haven't taken the time to understand the algo so this could be a misguided comment.

I don't think Bresenham has anything to do with what I need here.
This part of the algorithm is basically:
- get the equations of both lines delimiting the Arc/Sector
- for each point of the Circle, check if it is on the required side of both lines to decide if it is part of the Arc or not.

> 4. Are all the tiny `combine_*` functions required?

They are actually not strictly required. I did it this way to avoid having lots of comparisons / logic in the hot path (ie. pixel iterator). So basically I decide the logic to apply while constructing the iterator by just choosing which of the `combine_*` function to use, and then I can simply call the appropriate `combine_*` logic in the iterator itself.
But honestly, that might be premature optimization. I could try to move all the logic into the iterator.

jamwaffles commented:
> But I have an issue with your (center: Point, start: Point, end: Point) definition. It looks over-constrained to me. What if the distance between center and start is not the same as the distance between center and stop ? We couldn't draw a Circle or Arc that passes thru both start and stop.

That's a great point. I've added a comment [here](https://github.com/jamwaffles/embedded-graphics/issues/284#issuecomment-609991263) to discuss it further. I think this PR may be a bit premature as there are some bits of #284 that haven't been finalised yet.

aurelj commented:
Here is a much improved version of this patch set.

It now has two separate Primitives for `Arc` and `Sector` (`Arc` having no fill). There is some code duplication between the two, but we can get rid of this later on.

I replaced my `FixedPoint` implementation using the `fixed` crate.

I also got rid of the `combine_*` function and split the algorithm into two separate Iterator, one that return the points of the bounding box that are in the desired sector of the plane, and one that return the distance from each of this points to the center (in the same spirit as the new `DistanceIterator` in `Circle`).

Now I have a question regarding `Arc`. As it is not a closed shape and it has no surface (except for the stroke width), I guess it should not implement `ContainsPoint`, but I wonder if its `Points` iterator should return the points covered by the stroke width or nothing at all ? The documentation for `PointsIter` state "all points inside the primitive", but since `Arc` has no inside (ie. it is not a closed shape), I choose to not return any point. Let me know if I should change this.

Overall, I think this patch set is in a pretty good shape now.

rfuest commented:
I haven't had time to look at the code, but I can at least answer your question.

> Now I have a question regarding `Arc`. As it is not a closed shape and it has no surface (except for the stroke width), I guess it should not implement `ContainsPoint`, but I wonder if its `Points` iterator should return the points covered by the stroke width or nothing at all ? The documentation for `PointsIter` state "all points inside the primitive", but since `Arc` has no inside (ie. it is not a closed shape), I choose to not return any point. Let me know if I should change this.

You are correct that `CointainsPoint' should not be implemented for `Arc`.

The `PointsIter` docs need an update to mention non closed shapes. For non closed shapes `PointsIter` behaves differently by iterating over all points along the shape. This means that it should return the same points as the corresponding styled iter when drawing a 1px wide stroke.

aurelj commented:
> The `PointsIter` docs need an update to mention non closed shapes. For non closed shapes `PointsIter` behaves differently by iterating over all points along the shape. This means that it should return the same points as the corresponding styled iter when drawing a 1px wide stroke.

Oh, I see. I've now implemented `PointsIter` for `Arc` to match this definition.

I think this PR is now ready for review.

rfuest commented:
Before reviewing this we should decide how e-g should handle angles in general. Fixed point arithmetic could be beneficial on some architectures, but because microcontrollers with floating point units are getting more common I would also like to support floating point angles. We will need to decide if we want to support both or if one is enough.

@aurelj: Do you plan to use e-g on a target platform that would greatly benefit from using the fixed point implementation? Or could you also use something like the `micromath` crate with `f32` angles without a huge impact on performance or code size?

aurelj commented:
First note that in my implementation, `Angle` does not expose any floating or fixed point in its public API. It is only used internally, so it is easy to change at any time.
I guess it shouldn't be too hard to have 2 different internal implementations that can be chosen at build time with a feature flag. That would help comparing both implementations on different platforms. I will give it a try.

I'm working on various STM32 platforms. For now I'm doing some e-g testing on a STM32F4 which does support hardware floating point, but many of the popular STM32 don't have a FPU (STM32F0, STM32F1 and STM32F2). This includes the Blue Pill (STM32F103) which is one of the most common beginner platform for embedded-rust.

I don't want to be "forced" to choose a high end MCU when using e-g just because of the choice of using floating point for some internal computation. So I think if we choose to only have the floating point implementation, it first has to be seriously evaluated on MCU without FPU first.

rfuest commented:
> First note that in my implementation, `Angle` does not expose any floating or fixed point in its public API. It is only used internally, so it is easy to change at any time.

But it also only allows integer degrees, which should be enough resolution for smaller displays but larger displays (>= 320x240 px) could benefit from < 1 degree steps.

Can we always use `f32` for the user facing API, even if we will use fixed point internally? I would prefer not to have different APIs when changing a feature.

> I don't want to be "forced" to choose a high end MCU when using e-g just because of the choice of using floating point for some internal computation. So I think if we choose to only have the floating point implementation, it first has to be seriously evaluated on MCU without FPU first.

Sure this needs to be considered and lower end MCUs need be supported well. But not using advanced features, if present, isn't great either.

@jamwaffles: What are your thoughts on this subject?

jamwaffles commented:
> Can we always use f32 for the user facing API, even if we will use fixed point internally?

This would be ideal yeah, even if the first thing that happens is to convert the `f32` to fixed point. It gives us the freedom to redefine the internal implementation.

Unfortunately `micromath` doesn't have any benchmarks but it might be a good compromise between angle accuracy and performance so I'd be leaning towards that, at least for an initial implementation.

Using `f32` and `micromath` I think would allow the removal of `Angle` and the associated boilerplate as well?

rfuest commented:
> Using `f32` and `micromath` I think would allow the removal of `Angle` and the associated boilerplate as well?

Keeping `Angle` provides some type safety for mixing radians and degrees.
An extension trait for `f32` it could make `Angle` nicer to use: `90.0.deg()`, `3.141.rad()`

jamwaffles commented:
> Keeping Angle provides some type safety for mixing radians and degrees.

True, that sounds like a useful feature.

Shall we settle on `micromath` and a trait on `f32` that adds `.angle()` for now? I'm not sure if we need to explore anything else at the moment. Radian support and faster/better internal implementations can always be added later on.

rfuest commented:
> Shall we settle on `micromath` [...] for now?

If @aurelj agrees I think that's a good solution.

>  [...] and a trait on `f32` that adds `.angle()` for now?

IMO `.angle()` doesn't make much sense, because it doesn't define the unit of the value. I would suggest to use `fn rad(self) -> Angle` and `fn deg(self) -> Angle`, where `rad` is a basic constructor and `deg` converts the given value into radians (the internal unit).

jamwaffles commented:
> IMO .angle() doesn't make much sense

Argh, yes you're right! I was flying through issues too fast between things at work. I meant to say `.deg()` and `.rad()` as you originally suggested :facepalm:

aurelj commented:
OK, here is a new revision with, I think, all of your concerns taken into account.

The `Angle` facing API is now using `f32` to support higer precision. It also support specifying angles in degree or radian. They are stored internally as radian.
I also added an `f32` extension trait to be able to use 90.0.deg() or PI.rad().

Internally, all the values are stored in a `Real` wrapper type (better name ?). `Real` is implemented as either `f32` or `I16F16` fixed point from the `fixed` crate. To use the fixed point version, you need to use `cargo --features fixed_point`. By default the `f32` implementation is used.

The trigonometric functions are implemented using micromath for the `f32` version and using a LUT for the `I16F16` version.

Let me know if this matches a bit better what you had in mind.

aurelj commented:
I've just fixed the `float-cmp` crate so that it builds with no_std.
Until the next `float-cmp` release, I've pointed the dependency to upstream git.

aurelj commented:
And... `float-cmp` has a new release, so no more dependency on the git master.

rfuest commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,634 @@
+
+impl PlaneSector {
+    fn new(arc: &Arc) -> Self {
+        let angle_start = arc.angle_start.normalize_from(-90.0.deg());
+        let angle_end = arc.angle_end().normalize_from(-90.0.deg());
```
Why does this subtract 90 degree from the angle?

aurelj replied:
First note that it doesn't subtract 90 degree from the angle, it selects the normalization starting point to -90 rather than 0.

Basically, to decide whether we should draw on the part of the plane which is below or above line_a (variable draw_above_a), we need to know if the angle is either between -90 and 90, or if it is between 90 and 270. To do this, the easiest way is to normalize the angle between -90 and 270 rather than 0 to 360.

rfuest commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,634 @@
+        Self {
+            line_a: LinearEquation::from_point_angle(arc.center(), arc.angle_start),
+            line_b: LinearEquation::from_point_angle(arc.center(), arc.angle_end()),
+            draw_above_a: (angle_start < 90.0.deg()) ^ negative_sweep,
+            draw_above_b: (angle_end >= 90.0.deg()) ^ negative_sweep,
```
Adding constants for `90.0.deg()` (and maybe other important angles) could reduce the number of floating point operations.

aurelj replied:
Yep, I kind of thought about adding some consts at some point without actually doing it...
Done now, for 90deg, 180deg and 360deg.
No more use of float.deg() conversion in the code base (except in tests and examples).

rfuest commented on embedded-graphics/src/primitives/mod.rs:
````diff
@@ -41,6 +44,111 @@ pub trait ContainsPoint {
+/// assert_eq!(arc_1, arc_2);
+/// ```
+#[macro_export]
+macro_rules! egarc {
````
IMO the macro shouldn't convert from degree to `Angle` internally, but take `Angle` as an argument instead. This way the user can use `180.0.deg()`, `PI.rad()` and even `some_angle_variable + 123.0.deg()` as angles.

aurelj replied:
Yes, you're right. I didn't put much thought about those macros after adding support for floating point angles and radiant. Changed to take `Angle` argument now.

rfuest commented on embedded-graphics/src/primitives/mod.rs:
````diff
@@ -387,6 +495,111 @@ macro_rules! egrectangle {
+/// assert_eq!(sector_1, sector_2);
+/// ```
+#[macro_export]
+macro_rules! egsector {
````
The note about angles for `egarc` also applies to this macro.

aurelj replied:
Done.

rfuest commented on simulator/examples/pacman.rs:
```diff
@@ -0,0 +1,68 @@
+    let mut display: SimulatorDisplay<Rgb565> = SimulatorDisplay::new(Size::new(65, 65));
+
+    // Create styles used by the drawing operations.
+    let sector_stroke = PrimitiveStyleBuilder::new()
```
These styles apply both stroke and fill:
```suggestion
    let sector_style = PrimitiveStyleBuilder::new()
```

aurelj replied:
OK.

rfuest commented on simulator/examples/pacman.rs:
```diff
@@ -0,0 +1,68 @@
+        .stroke_width(2)
+        .fill_color(Rgb565::YELLOW)
+        .build();
+    let eye_stroke = PrimitiveStyleBuilder::new()
```
```suggestion
    let eye_style = PrimitiveStyleBuilder::new()
```

aurelj replied:
OK.

rfuest commented on tinytga/src/lib.rs:
```diff
@@ -289,7 +289,7 @@ impl<'a> Iterator for TgaIterator<'a> {
         self.x += 1;
 
-        if self.x >= self.tga.width().into() {
+        if self.x >= self.tga.width() as u32 {
```
Please remove the changes to `tinytga` because they aren't related to this PR.

aurelj replied:
In fact the 3 changes to `tinytga` are related to this PR. When the `fixed` crate is used, some of those types can't be inferred anymore. Here is the error you get with this code with the `fixed` crate in use:

```
error[E0283]: type annotations needed for `u32`
   --> tinytga/src/lib.rs:292:19
    |
250 |         let mut pixel_value = {
    |             --------------- consider giving `pixel_value` a type
...
292 |         if self.x >= self.tga.width().into() {
    |                   ^^ cannot infer type for type `u32`
    |
    = note: cannot satisfy `u32: core::cmp::PartialOrd<_>`

error: aborting due to previous error
```

rfuest replied:
OK, interesting, I didn't expect that. Can you change this to:
```suggestion
        if self.x >= u32::from(self.tga.width()) {
```
which is a little safer than a cast if the return type of `tga.with` ever changes.

aurelj replied:
OK.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+
+    /// Normalize the angle to less than one full rotation, starting from angle.
+    /// (ie. in the range angle..(angle+360)).
+    pub fn normalize_from(self, angle: Self) -> Self {
```
Indeed, I agree. Done.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+            I16F16::from_bits(65526),
+            I16F16::from_bits(65536),
+        ];
+        let degree: i32 = (Real::from(180.0) * self.0 / Real::from(PI)).round().into();
```
Should be better now with added constant for PI and converting from int for 180.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+    }
+
+    fn cos(self) -> Real {
+        (self + 90.0.deg()).sin()
```
Using a constant now, so since this is fixed point code only, this is a simple integer addition.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+
+    fn tan(self) -> Option<Real> {
+        let cos = self.cos();
+        if cos != 0.0.into() {
```
Right. Done.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+    fn from(other: Angle) -> f32 {
+        other.0.into()
+    }
+}
```
Indeed. That's some remaining code from initial version which supported only integer degree that I converted to f32 without thinking about it. Removed now.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+    }
+
+    #[test]
+    fn sin_correct() {
```
Yes, good idea. That looks much better.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+    }
+
+    #[test]
+    fn cos_correct() {
```
Done.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+    }
+
+    #[test]
+    fn tan_correct() {
```
Done.

aurelj commented on embedded-graphics/src/geometry/real.rs:
````diff
@@ -0,0 +1,206 @@
+/// point depending on feature flag.
+/// ```
+
+#[cfg(not(feature = "fixed_point"))]
````
Good idea as well. This looks cleaner now with those private mod.

aurelj commented on embedded-graphics/src/geometry/real.rs:
```diff
@@ -0,0 +1,206 @@
+    }
+}
+
+use float_cmp::{ApproxEq, F32Margin};
```
Oh, my mistake. Put it there at first for a quick test of float_cmp and forgot to move it up when I finally seriously adopted float_cmp.

aurelj commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,634 @@
+            }
+        }
+
+        None
```
I like it better this way indeed.

aurelj commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,634 @@
+    outer_threshold: u32,
+    inner_threshold: u32,
+
+    color: Option<C>,
```
Ok, fine with me.

aurelj commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,634 @@
+        }
+
+        None
+    }
```
Right, I like it.

aurelj commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,636 @@
+    fn tan(self) -> Option<Real> {
+        let angle: f32 = self.0.into();
+        let tan = angle.tan();
+        if tan.is_nan() || tan.abs() > 20000000.0 {
```
tan(90) has no value (tending toward INFINITY...). So we need a return value for this case, which works with Real whether it is implemented as floating point or fixed point. That's why I chose to return `Option<Real>` so tan(90) can return `None`.

This `None` return value is then used in `LinearEquation` to generate the equation of a vertical line (in the form of `x + c = 0`).

The issue here is the approximation returned by `micromath` for 90.0.tan() (a big value slightly greater than 20000000.0). So to be consistent with the fixed point code, I chose to detect when tan() is huge, and consider that it means that the angle is (+/-)90 degree and thus consider that its tan() has no actual numerical value.

rfuest replied:
OK, thanks for the explanation. The main reason for the question was to know how the magic value is derived. I'll take another look at this when I have looked at the drawing algorithm.

Can you add a comment to the code that describes why the value `20000000.0` was chosen?

aurelj replied:
Yes, adding comment now.

jamwaffles commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,797 @@
+        let inner_threshold = circle::diameter_to_threshold(fill_area.diameter);
+        let outer_threshold = circle::diameter_to_threshold(stroke_area.diameter);
+
+        let line_a = stroke_area.line_from_angle(styled.primitive.angle_start);
+        let line_b = stroke_area.line_from_angle(styled.primitive.angle_end());
```
Won't this draw pixels near the center of the arc multiple times? I'm ok to leave it as is for this PR, but we should maybe open an issue to optimise arc filling. Also, other primitives AFAIK make an effort to not draw any two pixels twice so it would be good to have `Sector` mirror this behaviour.

jamwaffles commented:
This is looking really promising. The demo is cool!

But... I found a small bug with `Sector` with a stroke:

![image](https://user-images.githubusercontent.com/800791/81110210-23714d80-8f13-11ea-9114-42584f49592e.png)

The bottom green line of the sector is 1px too long. In the code below, if I set `progress` to `35` the line renders correctly.

It also looks like the vertical starting line is off by one also. Is the sector too small, or the line too long maybe?

Code to reproduce:

```rust
let progress = 34;

let sweep = progress as f32 * 360.0 / 100.0;

Sector::new(Point::new(16, 16), 64 - 32, 90.0.deg(), sweep.deg())
    .into_styled(
        PrimitiveStyleBuilder::new()
            .stroke_color(Rgb565::GREEN)
            .fill_color(Rgb565::RED)
            .stroke_width(1)
            .stroke_alignment(StrokeAlignment::Inside)
            .build(),
    )
    .draw(&mut display)?;
```

The effect is a bit more pronounced with a stroke width of `5`:

![image](https://user-images.githubusercontent.com/800791/81110753-e0fc4080-8f13-11ea-8f38-e878d16cd835.png)

aurelj commented:
> This is looking really promising. The demo is cool!
> 
> But... I found a small bug with `Sector` with a stroke:
> 
> [...]
> 
> The bottom green line of the sector is 1px too long.

Oh, right. There seems to be some kind of off by one in the length of the 2 stroke lines delimiting the sector. And this is actually also visible in the pacman example. I think it wasn't present in some previous version of this patchset.

I guess this is related to some changes I made to accommodate the new .stroke_alignment() property recently, but off the top of my head I'm not sure what. I will try to have a look at it.

aurelj commented:
> > This is looking really promising. The demo is cool!
> > But... I found a small bug with `Sector` with a stroke:
> > [...]
> > The bottom green line of the sector is 1px too long.
> 
> Oh, right. There seems to be some kind of off by one in the length of the 2 stroke lines delimiting the sector. And this is actually also visible in the pacman example. I think it wasn't present in some previous version of this patchset.

This was a simple rounding issue when computing the `Point` at the end of the `Line`.
Fixed now.

jamwaffles commented:
Just gave this another test. The end line is indeed fixed, but the vertical start looks to be 1px off still:

![image](https://user-images.githubusercontent.com/800791/81910334-b9684080-95c3-11ea-83ec-c79dcfdc088e.png)

The arc has a radius of 16px but the line is 17px long.

aurelj commented:
> Just gave this another test. The end line is indeed fixed, but the vertical start looks to be 1px off still:

Fixed now.

aurelj commented:
> 1. `lib.rs` and `embedded-graphics/README.md` both have a list of primitives available. Can you add `Arc` and `Sector` to them please?

OK, done.

> 2. I've started adding benchmarks at the beginning of new primitive development to `simulator/benches/primitives.rs` so we can track performance over time.

Oh, I didn't notice this. Nice !

> Can you add one or two for `Arc` and `Sector` to that suite?

Sure, done.

> It might also be worth adding a separate bench for the `fixed_point` feature toggle with something like this in `Cargo.toml` [...]

Hum... OK, I tried to add this but I'm not sure it is such a great thing. With my last commit, you can now use `cargo bench --features fixed_point` which will indeed run the benches that are in `primitive_fixed_point.rs`, but it will also run all the other benches (including the identical `primitive.rs` benches) compiled with a fixed_point version of e-g. So I'm not sure if it is useful to have a specific `primitive_fixed_point.rs`, and it may even be a good thing to test everything with fixed_point, just in case a regression is introduced later on, in some code that wasn't supposed to make use of fixed_point...

rfuest commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,632 @@
+}
+
+impl Arc {
+    /// Create a new arc delimited with a top-left point with a specific diamter and start and sweep angles
```
```suggestion
    /// Create a new arc delimited with a top-left point with a specific diameter and start and sweep angles
```

rfuest commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,632 @@
+        angle_start: Angle,
+        angle_sweep: Angle,
+    ) -> Self {
+        let top_left = center - Size::new(diameter, diameter).center_offset();
```
```suggestion
        let top_left = center - Size::new_equal(diameter).center_offset();
```

rfuest commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,632 @@
+
+/// Iterator that returns the squared distance to the center for all points in the bounding box.
+#[derive(Copy, Clone, PartialEq, PartialOrd, Debug)]
+struct DistanceIterator {
```
We should merge this with `circle::DistanceIter`, by replacing `circle::DistanceIter` with:
(I didn't test the code, but it hope it's working enough to describe what I mean)
```rust
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
pub(in crate::primitives) struct DistanceIterator<I> {
    center: Point,
    points: I,
}

impl<I> DistanceIterator<I>
where
    I: Iterator<Item = Point>
{
    pub(in crate::primitives) fn new(center: Point, points: I) -> Self {
        Self {
            center,
            points,
        }
    }
}

impl<I> Iterator for DistanceIterator<I>
where
    I: Iterator<Item = Point>
{
    type Item = (Point, u32);

    fn next(&mut self) -> Option<Self::Item> {
        self.points.next().map(|p| {
            let delta = self.center - p * 2;
            let distance = delta.x.pow(2) as u32 + delta.y.pow(2) as u32;

            (p, distance)
        })
    }
}
```
Which could then be used for `Circle`, `Arc` and `Sector`.

rfuest commented on embedded-graphics/src/primitives/arc.rs:
```diff
@@ -0,0 +1,632 @@
+
+/// Define one side of a line
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+enum LineSide {
+    ABOVE,
+    BELLOW,
+}
```
Enum variants need to be camel case and there is a typo:
```suggestion
enum LineSide {
    Above,
    Below,
}
```

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+impl ContainsPoint for Sector {
+    fn contains(&self, point: Point) -> bool {
+        let delta = self.center_2x() - point * 2;
+        let distance = delta.x.pow(2) as u32 + delta.y.pow(2) as u32;
```
After rebasing you can use the new: `delta.length_squared() as u32`.

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+            return false;
+        }
+
+        let angle_start = self.angle_start.normalize_from(-ANGLE_90DEG);
+        let angle_end = self.angle_end().normalize_from(-ANGLE_90DEG);
+        let negative_sweep = self.angle_sweep < Angle::zero();
+
+        let line_a = LinearEquation::from_point_angle(self.center(), self.angle_start);
+        let line_b = LinearEquation::from_point_angle(self.center(), self.angle_end());
+        let draw_above_a = (angle_start < ANGLE_90DEG) ^ negative_sweep;
+        let draw_above_b = (angle_end >= ANGLE_90DEG) ^ negative_sweep;
+        let sweep = self.angle_sweep.abs();
+
+        let side_a = line_a.side(&point);
+        let side_b = line_b.side(&point);
```
Couldn't you instead use `PlaneSector::new`?

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+
+impl Dimensions for Sector {
+    fn bounding_box(&self) -> Rectangle {
+        Rectangle::new(self.top_left, Size::new(self.diameter, self.diameter))
```
```suggestion
        Rectangle::new(self.top_left, Size::new_equal(self.diameter))
```

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+enum IterState {
+    ARC,
+    LINES,
+    DONE,
+}
```
Enum variants need to be camel case:
```suggestion
enum IterState {
    Arc,
    Lines,
    Done,
}
```

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+        let side_a = line_a.side(&point);
+        let side_b = line_b.side(&point);
+
+        let correct_a_side = draw_above_a ^ (side_a == LineSide::BELLOW);
+        let correct_b_side = draw_above_b ^ (side_b == LineSide::BELLOW);
+
+        if sweep < ANGLE_180DEG {
+            correct_a_side && correct_b_side
+        } else if sweep < ANGLE_360DEG {
+            correct_a_side || correct_b_side
+        } else {
+            true
+        }
+    }
```
This is the same code used in `PlaneSectorIterator::next`. I would suggest to add this to `PlaneSector`:
```rust
fn contains(&self, point: Point) -> bool {
    let side_a = self.line_a.side(point);
    let side_b = self.line_b.side(point);

    let correct_a_side = self.draw_above_a ^ (side_a == LineSide::BELLOW);
    let correct_b_side = self.draw_above_b ^ (side_b == LineSide::BELLOW);

    if self.sweep < ANGLE_180DEG {
        correct_a_side && correct_b_side
    } else if self.sweep < ANGLE_360DEG {
        correct_a_side || correct_b_side
    } else {
        true
    }
}
```
And use it in `PlaneSectorIterator::next` and `ContainsPoint::contains`.

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+}
+
+#[derive(Copy, Clone, PartialEq, PartialOrd, Debug)]
+struct PlaneSector {
```
Am I right that this identical to `arc::PlaneSector`? If it is, it shouldn't be duplicated but imported.

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+
+/// Iterator that returns only the points which are inside a plane sector defined by two lines.
+#[derive(Copy, Clone, PartialEq, PartialOrd, Debug)]
+struct PlaneSectorIterator {
```
See comment for `PlaneSector`.

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+
+/// Iterator that returns the squared distance to the center for all points in the bounding box.
+#[derive(Copy, Clone, PartialEq, PartialOrd, Debug)]
+struct DistanceIterator {
```
See comment for `arc::DistanceIterator`.

rfuest commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,798 @@
+///
+/// The equation is stored as the a, b and c coefficients of the ax + by + c = 0 equation
+#[derive(Copy, Clone, PartialEq, PartialOrd, Debug)]
+struct LinearEquation {
```
This also seems to be identical to `arc::LinearEquation`.

jamwaffles commented:
> If you like you could also apply the new code layout for primitives to Arc and Sector

I'm happy to tackle this over in #361 so we can land this PR sooner. It's been open quite a while and would be great to get merged in.

jamwaffles commented on embedded-graphics/README.md:
```diff
@@ -18,6 +18,8 @@ It contains built in items that make it easy to draw 2D graphics primitives:
   - Rectangles
   - Circles
   - Ellipses
+  - Arc
+  - Sector
```
Pedantic nit:

```suggestion
  - Arcs
  - Sectors
```

jamwaffles commented on embedded-graphics/src/lib.rs:
```diff
@@ -7,6 +7,8 @@
 //!     * [Rectangles (and squares)](./primitives/rectangle/struct.Rectangle.html)
 //!     * [Circles](./primitives/circle/struct.Circle.html)
 //!     * [Ellipses](./primitives/ellipse/struct.Ellipse.html)
+//!     * [Arc](./primitives/arc/struct.Arc.html)
+//!     * [Sector](./primitives/sector/struct.Sector.html)
```
```suggestion
//!     * [Arcs](./primitives/arc/struct.Arc.html)
//!     * [Sectors](./primitives/sector/struct.Sector.html)
```

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,477 @@
+    }
+
+    /// Return numerical value of the angle in degree
+    pub fn to_degree(self) -> f32 {
```
```suggestion
    pub fn to_degrees(self) -> f32 {
```

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,477 @@
+    }
+
+    /// Return numerical value of the angle in radian
+    pub fn to_radian(self) -> f32 {
```
```suggestion
    pub fn to_radians(self) -> f32 {
```

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
````diff
@@ -0,0 +1,477 @@
+///
+/// ## Create an `Angle` from a value using `AngleUnit`
+///
+/// ```rust
+/// use embedded_graphics::geometry::AngleUnit;
+///
+/// // Create an angle using the `AngleUnit` method
+/// let angle_b = 30.0.deg();
````
```suggestion
/// let angle_b = 30.0.deg();
/// ```
```

jamwaffles replied:
For completeness' sake, it would be great if this example showed usage of `.rad()` as well.

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,477 @@
+/// // Create an angle using the `AngleUnit` method
+/// let angle_b = 30.0.deg();
+pub trait AngleUnit {
+    /// Convert a number (interpreted as degree) to an `Angle`.
```
```suggestion
    /// Convert a number (interpreted as degrees) to an `Angle`.
```

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,477 @@
+    /// Convert a number (interpreted as degree) to an `Angle`.
+    fn deg(self) -> Angle;
+
+    /// Convert a number (interpreted as radian) to an `Angle`.
```
```suggestion
    /// Convert a number (interpreted as radians) to an `Angle`.
```

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,477 @@
+        Angle((angle * PI / 180.0).into())
+    }
+
+    /// Creates an angle defined in radian.
+    pub fn from_radian(angle: f32) -> Self {
```
```suggestion
    /// Creates an angle defined in radians.
    pub fn from_radians(angle: f32) -> Self {
```

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,477 @@
+
+impl Angle {
+    /// Creates an angle defined in degrees.
+    pub fn from_degree(angle: f32) -> Self {
```
```suggestion
    pub fn from_degrees(angle: f32) -> Self {
```

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,477 @@
+    }
+}
+
+/// AngleUnit trait.
```
I think this should be reexported in `prelude.rs` unless @rfuest disagrees.

rfuest replied:
No, I don't disagree.

jamwaffles replied:
Ah so it is, my bad.

jamwaffles commented on embedded-graphics/src/geometry/angle.rs:
```diff
@@ -0,0 +1,477 @@
+/// use embedded_graphics::geometry::Angle;
+///
+/// // Create an angle using the `from_degree` constructor method
+/// let angle_a = Angle::from_degree(10.0);
```
Similar to my other comment, it would be useful to show `Angle::from_radian(s)` here as well.

Another example showing usage of the `AngleUnit` (with a doc link to it as well) would be handy too.

jamwaffles commented on embedded-graphics/src/geometry/real.rs:
```diff
@@ -0,0 +1,214 @@
+
+/// Real.
+///
+/// `Real` is used to store real numbers as either floating point or fixed
+/// point depending on feature flag.
```
This should mention the feature in question:

```suggestion
/// `Real` is used to store real numbers as either floating point or 32 bit fixed point if the
/// `fixed_point` cargo feature is enabled.
```

aurelj commented:
> Sorry for the long delay in reviewing this, I would like to get this merged soon. Can you rebase the code one more time? The macros were recently removed and `DrawTarget` doesn't need a `draw_arc` and `draw_sector` method anymore.

Rebased against latest master, with all the small needed adaptations.

> If you like you could also apply the new code layout for primitives to `Arc` and `Sector`, that @jamwaffles is currently working on for the other primitives: #363. But this can also be done after this PR is merged.

@jamwaffles, if you don't mind, I would happily let you split the `Arc` and `Sector` primitives into multiple files.

aurelj commented:
> Please also add the `fixed_point` feature with a short description to the features list in both `lib.rs` and `README.md`.

Done as well.

jamwaffles commented on embedded-graphics/CHANGELOG.md:
```diff
@@ -27,6 +27,10 @@ Embedded Graphics is a `no_std` library for adding graphics features to display
 - [#366](https://github.com/jamwaffles/embedded-graphics/pull/366) Allow usage of all grayscale color types in `MockDisplay` patterns.
 - [#342](https://github.com/jamwaffles/embedded-graphics/pull/342) Added `Rectangle::intersection` method.
 - [#363](https://github.com/jamwaffles/embedded-graphics/pull/363) Export `primitives::ContainsPoint` trait in prelude.
+- [#310](https://github.com/jamwaffles/embedded-graphics/pull/320) Added the `Arc` primitive.
+- [#310](https://github.com/jamwaffles/embedded-graphics/pull/320) Added `DrawTarget::draw_arc`.
+- [#310](https://github.com/jamwaffles/embedded-graphics/pull/320) Added the `Sector` primitive.
+- [#310](https://github.com/jamwaffles/embedded-graphics/pull/320) Added `DrawTarget::draw_sector`.
```
The `DrawTarget` methods are not longer added

```suggestion
- [#310](https://github.com/jamwaffles/embedded-graphics/pull/320) Added the `Arc` primitive.
- [#310](https://github.com/jamwaffles/embedded-graphics/pull/320) Added the `Sector` primitive.
```

aurelj replied:
Oh, of course... Done.

jamwaffles commented on embedded-graphics/src/primitives/sector.rs:
```diff
@@ -0,0 +1,646 @@
+}
+
+impl Sector {
+    /// Create a new sector delimited with a top-left point with a specific diamter and start and sweep angles
```
```suggestion
    /// Create a new sector delimited with a top-left point with a specific diameter and start and sweep angles
```

aurelj replied:
Fixed.
